esp-config/src/generate.rs

```diff
@@ -0,0 +1,336 @@
+        *self = match self {
+            Value::Bool(_) => match s {
+                "false" | "no" | "n" => Value::Bool(false),
+                "true" | "yes" | "y" => Value::Bool(true),
+                _ => return Err(ParseError(format!("Invalid boolean value: {}", s))),
+            },
+            Value::Number(_) => Value::Number(
+                match s.as_bytes() {
+                    [b'0', b'x', ..] => usize::from_str_radix(&s[2..], 16),
+                    [b'0', b'o', ..] => usize::from_str_radix(&s[2..], 8),
+                    [b'0', b'b', ..] => usize::from_str_radix(&s[2..], 2),
+                    _ => usize::from_str_radix(&s, 10),
+                }
+                .map_err(|_| ParseError(format!("Invalid numerical value: {}", s)))?,
+            ),
+            Value::String(_) => Value::String(String::from(s)),
+        };
+        Ok(())
+    }
+
+    fn as_string(&self) -> String {
+        match self {
+            Value::Bool(value) => String::from(if *value { "true" } else { "false" }),
+            Value::Number(value) => format!("{}", value),
+            Value::String(value) => value.clone(),
+        }
+    }
+}
+
+/// Generate and parse config from a prefix, and array of key, default,
+/// description tuples.
+///
+/// This function will parse any `SCREAMING_SNAKE_CASE` environment variables
+/// that match the given prefix. It will then attempt to parse the [`Value`].
+/// Once the config has been parsed, this function will emit `snake_case` cfg's
+/// _without_ the prefix which can be used in the dependant crate. After that,
+/// it will create a markdown table in the `OUT_DIR` under the name
+/// `{prefix}_config_table.md` where prefix has also been converted
+/// to `snake_case`. This can be included in crate documentation to outline the
+/// available configuration options for the crate.
+///
+/// Passing a value of true for the `emit_md_tables` argument will create and
+/// write markdown files of the available configuration and selected
+/// configuration which can be included in documentation.
+///
+/// Unknown keys with the supplied prefix will cause this function to panic.
+pub fn generate_config(
+    prefix: &str,
+    config: &[(&str, Value, &str)],
+    emit_md_tables: bool,
+) -> HashMap<String, Value> {
+    // only rebuild if build.rs changed. Otherwise Cargo will rebuild if any
+    // other file changed.
+    println!("cargo:rerun-if-changed=build.rs");
+
+    // ensure that the prefix is `SCREAMING_SNAKE_CASE`
+    let mut prefix = screaming_snake_case(prefix);
+    let mut doc_table = String::from(DOC_TABLE_HEADER);
+    let mut selected_config = String::from(CHOSEN_TABLE_HEADER);
+
+    let mut configs = create_config(&prefix, config, &mut doc_table);
+    capture_from_env(&prefix, &mut configs);
+    emit_configuration(&prefix, &configs, &mut selected_config);
+
+    if emit_md_tables {
+        // convert to snake case for the file name
+        prefix.make_ascii_lowercase();
```

Comment:
If you used `snake_case` here and placed the result in `file_name` you could drop the comment I think :)